docs/content/docs/dev/table/data_stream_api.md

```diff
@@ -0,0 +1,2106 @@
+// | -D |                          Alice |          12 |
+// | +I |                          Alice |         100 |
+// +----+--------------------------------+-------------+
+
+
+// === EXAMPLE 2 ===
+
+// interpret the stream as an upsert stream (without a need for UPDATE_BEFORE)
+
+// create a changelog DataStream
+val dataStream = env.fromElements(
+    Row.ofKind(RowKind.INSERT, "Alice", Int.box(12)),
+    Row.ofKind(RowKind.INSERT, "Bob", Int.box(5)),
+    Row.ofKind(RowKind.UPDATE_AFTER, "Alice", Int.box(100))
+)(Types.ROW(Types.STRING, Types.INT))
+
+// interpret the DataStream as a Table
+val table =
+    tableEnv.fromChangelogStream(
+        dataStream,
+        Schema.newBuilder().primaryKey("f0").build(),
+        ChangelogMode.upsert())
+
+// register the table under a name and perform an aggregation
+tableEnv.createTemporaryView("InputTable", table)
+tableEnv
+    .executeSql("SELECT f0 AS name, SUM(f1) AS score FROM InputTable GROUP BY f0")
+    .print()
+
+// prints:
+// +----+--------------------------------+-------------+
+// | op |                           name |       score |
+// +----+--------------------------------+-------------+
+// | +I |                            Bob |           5 |
+// | +I |                          Alice |          12 |
+// | -D |                          Alice |          12 |
+// | +I |                          Alice |         100 |
+// +----+--------------------------------+-------------+
+```
+{{< /tab >}}
+{{< /tabs >}}
+
+The default `ChangelogMode` shown in example 1 should be sufficient for most use cases as it accepts
+all kinds of changes.
+
+However, example 2 shows how to limit the kinds of incoming changes for efficiency by reducing the
+number of update messages by 50%.
+
+### Examples for `toChangelogStream`
+
+The following code shows how to use `toChangelogStream` for different scenarios.
+
+{{< tabs "fc4cd538-4345-49ee-b86e-b308f002e069" >}}
+{{< tab "Java" >}}
+```java
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.functions.ProcessFunction;
+import org.apache.flink.table.api.DataTypes;
+import org.apache.flink.table.api.Schema;
+import org.apache.flink.table.api.Table;
+import org.apache.flink.table.data.StringData;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.Collector;
+import static org.apache.flink.table.api.Expressions.*;
+
+// create Table with event-time
+tableEnv.executeSql(
+    "CREATE TABLE GeneratedTable "
+    + "("
+    + "  name STRING,"
+    + "  score INT,"
+    + "  event_time TIMESTAMP_LTZ(3),"
+    + "  WATERMARK FOR event_time AS event_time - INTERVAL '10' SECOND"
+    + ")"
+    + "WITH ('connector'='datagen')");
+
+Table table = tableEnv.from("GeneratedTable");
+
+
+// === EXAMPLE 1 ===
+
+// convert to DataStream in the simplest and most general way possible (no event-time)
+
+Table simpleTable = tableEnv
+    .fromValues(row("Alice", 12), row("Alice", 2), row("Bob", 12))
```

Comment:
I noticed that `fromValues` actually doesn't necessarily create a source stream with the order as specified, so you have to be careful in the examples with that. For a `sum` as here, it doesn't matter though.
I don't think there's a better Table API alternative for starting an example though, except for maybe a "SELECT" statement with the values which doesn't seem to fit that well...

Reply:
A `SELECT` statement with values behaves the same. We still need a `fromElements` or `fromCollection` in Table API that keeps the order and allows any data type. Also rows with kinds can not be inserted yet.